### PR TITLE
[None][fix] size FP4 block-scale MoE permute buffers from local_num_experts

### PR DESCRIPTION
**Pull request by reasonsolo**
Split out of #19183 at @BowenFu's request.

## What

`run_fp4_block_scale_moe_runner` pads its permute buffers for every **global** expert, but the expert-parallel GEMMs below only materialize rows for the experts this rank owns — every GEMM weight is checked against `local_num_experts`. Padding for the global count over-allocates by up to the EP size, which can exhaust memory on large-token prefill shapes.

Passes `local_num_experts` to `getMaxPermutedPaddedCount` and `getMaxNumCtasInBatchDim`.

## Why it is its own PR

This is the FP4 block-scale path **shipping today on Blackwell**, so it shrinks a live allocation for every existing EP deployment. If the bound is ever short the failure mode is an out-of-bounds permute write rather than a clean error. It was originally carried in #19183 (Rubin SM107 enablement), where nothing exercises it — so it is split out to get a real build, an EP-sized run, and routing-kernel owner review on its own merits.

## Status / what reviewers should demand

- [x] Needs an EP-sized run (EP > 1) to confirm the tighter bound holds for the permute write.
- [x] Routing-kernel owner review on whether `getMaxPermutedPaddedCount` / `getMaxNumCtasInBatchDim` are intended to take the local or global expert count.

## Dev Engineer Review

- FP4 block-scale MoE workspace sizing uses `local_num_experts` for padded-token and CTA bounds. This reduces over-allocation in expert-parallel execution.
- `routingIndicesWarpKernel` filters routing counts to local experts and uses rank-local indices for CTA and offset calculations. Remote tokens retain `-1` permutation indices.
- No public API or configuration changes were identified. No material review findings were supplied.

## QA Engineer Review

- Added `LocalExpertCountBoundsPermuteAndCtaBuffers` and `LocalExpertCountNeverExceedsGlobalExpertCount` in `blockScaleMoeActivationTest.cu`.
- Added `WarpLevelParallelizationWithExpertParallelization` and `WarpLevelParallelizationNarrowLocalExpertWindow` in `routingLlama4Test.cpp`.
- Tests cover exhaustive local routing distributions, CTA and padded-row bounds, expert-parallel routing, and narrow local expert ownership.
- No test-list changes or registrations were identified.
- CI reported one successful run and one successful L0 pipeline. Coverage: **sufficient**.

## Per-File QA Perspective

- `RoutingLlama4.cu`: Verify local-expert filtering, rank-local CTA assignment, offset calculations, and `-1` handling for remote tokens.
- `fp4BlockScaleMoe.cpp`: Verify workspace allocation remains sufficient for local routing and decreases with expert parallelism.
- `blockScaleMoeActivationTest.cu`: Covers local routing capacity bounds and global-versus-local sizing invariants. No test-list registration was identified.
- `routingLlama4Test.cpp`: Covers Llama4 top-1 routing with expert parallelism and a narrow local expert window. No test-list registration was identified.

### CODE REVIEW

File: cpp/tensorrt_llm/thop/fp4BlockScaleMoe.cpp [hunk 1]
```diff
@@ -195,9 +195,12 @@ std::vector<torch::Tensor> run_fp4_block_scale_moe_runner(torch::optional<torch:
     auto routing_device = routing_logits.has_value() ? routing_logits.value().device() : topk_ids.value().device();
     at::Tensor num_tokens_per_expert
         = at::detail::empty_cuda({num_experts}, at::ScalarType::Int, routing_device, std::nullopt);
+    // The routing domain uses global expert IDs, but expert-parallel GEMMs only materialize rows for the experts owned
+    // by this rank. Padding intermediate buffers for every global expert over-allocates by up to the EP size, which can
+    // exhaust memory for large-token prefill shapes.
     int32_t max_num_padded_tokens
```
Comment (BowenFu) [inline cpp/tensorrt_llm/thop/fp4BlockScaleMoe.cpp:201]: The local-expert bound matches the GEMMs below, but this is also the out-of-bounds safety boundary for the shipped FP4 path, and the PR currently says the change has not been compiled or exercised. Would it be worth adding and running a focused EP>1 regression with uneven routing that proves these smaller permute and CTA buffers remain in bounds? This is required for this PR.

**Review by leslie-fang25 — APPROVED** [review]

File: cpp/tensorrt_llm/thop/moe/trtllmGen/fp4BlockScaleMoe.cpp [hunk 2]
```diff
@@ -195,9 +195,12 @@ std::vector<torch::Tensor> run_fp4_block_scale_moe_runner(torch::optional<torch:
     auto routing_device = routing_logits.has_value() ? routing_logits.value().device() : topk_ids.value().device();
     at::Tensor num_tokens_per_expert
         = at::detail::empty_cuda({num_experts}, at::ScalarType::Int, routing_device, std::nullopt);
+    // The routing domain uses global expert IDs, but expert-parallel GEMMs only materialize rows for the experts owned
+    // by this rank. Padding intermediate buffers for every global expert over-allocates by up to the EP size, which can
+    // exhaust memory for large-token prefill shapes.
     int32_t max_num_padded_tokens
```
Comment (reasonsolo) [reply cpp/tensorrt_llm/thop/moe/trtllmGen/fp4BlockScaleMoe.cpp:201]: Chasing the bound turned up a defect rather than a missing test — `routingIndicesWarpKernel` counted every token toward CTA capacity regardless of ownership and stored the *global* expert id in `mPtrCtaIdxXyToBatchIdx`, unlike every other routing path and unlike the suite's own host reference (`routingTest.cpp:173`/`:239`); `2c30eaa422` fixes it (gating only the capacity arithmetic, since `expertOffset` must stay ungated to keep the `-1` sentinel for remote-routed tokens) and `f83a58a475` adds the EP>1 Llama4 cases plus a host-side bound enumeration — none of it compiled or run here, so it needs a real `RoutingLlama4KernelTest` before merge.
